### PR TITLE
Adding GroupBy Support 

### PR DESCRIPTION
**Pull request by andrew-at-octopus**
Nevermore is based on a SQL database, but is missing some operations that can make queries simpler to write, and faster to run.
One missing feature is the GroupBy clause for obtaining aggregate information from the datastore.
While I was at it, I added a CountColumn helper method, which allows the (GroupBy) query to add a count column - mainly because the current Count() method expects that it finalises the query and is the only column returned (runs as scalar), whereas we want to include the count as one of the columns returned.

NOTE: Using GroupBy can easily result in queries that fall over, and at the moment there is no safeguard as to what columns are selected (eg, included in the group, or an aggregate), so be careful using it.

### CODE REVIEW

File: source/Nevermore/Querying/AST/Select.cs
Comment (emilol) [inline source/Nevermore/Querying/AST/Select.cs:30]: since `null` interpolates to empty string, could we simplify this to `var groupByString = groupBy?.GenerateSql()`

image

File: source/Nevermore/Querying/AST/Select.cs
Comment (emilol) [inline source/Nevermore/Querying/AST/Select.cs:12]: This class and its friends are public - does that mean they're part of our public API for nevermore? 

If so, as is this would be a breaking change, so we'll either need to add backwards compat, or do a major version bump.

File: source/Nevermore.Tests/QueryBuilderFixture/QueryBuilderFixture.cs
Comment (emilol) [inline source/Nevermore.Tests/QueryBuilderFixture/QueryBuilderFixture.cs:261]: Good to see test coverage, makes it super clear what we're adding 👍

File: source/Nevermore/Advanced/QueryBuilder.cs
Comment (emilol) [inline source/Nevermore/Advanced/QueryBuilder.cs:119]: we could reuse the overload that we're decorating here, to minimise our reasons to change.
```suggestion
            selectBuilder.CalculatedColumn("COUNT (*)", columnAlias);
```

File: source/Nevermore/Advanced/QueryBuilder.cs
Comment (emilol) [inline source/Nevermore/Advanced/QueryBuilder.cs:126]: as above
```suggestion
            CalculatedColumn($"COUNT ({distinctClause}{expression})", columnAlias);
```

**Review by emilol — APPROVED** [review]
LGTM - one key question about versioning/compat to sort out before merging, and a few minor nitpicks that shouldn't block this PR.

File: source/Nevermore/Querying/AST/Select.cs
Comment (andrew-at-octopus) [reply source/Nevermore/Querying/AST/Select.cs:30]: If I move the carriage return to the GenerateSql method yes.  Done

**Pull request merged**
